File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.py
from collections import deque


def solution(number, k):
    numbers = [i for i in number]
    a = deque()
    count = 0
    index = 1
    a.append(numbers[0])
    while count < k and index < len(numbers):
        temp = numbers[index]
        while count < k:
            if len(a) > 0 and a[len(a)-1] < temp:
                a.pop()
                count += 1
            else:
                a.append(temp)
                break
        index += 1
    return "".join((list(a) + numbers[index-1:])[:len(numbers)-1])

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.java
import java.util.*;

class Solution {
    public String solution(String number, int k) {
        Deque<Character> stack = new ArrayDeque<>();
        int count = 0; // 제거한 개수
        
        for (char digit : number.toCharArray()) {
            while (!stack.isEmpty() && count < k && stack.peekLast() < digit) {
                stack.pollLast(); 
                count++;
            }
            stack.addLast(digit);
        }
        
        StringBuilder result = new StringBuilder();
        int size = number.length() - k;
        for (int i = 0; i < size; i++) {
            result.append(stack.pollFirst());
        }
        
        return result.toString();
    }
}
